File: javascript/examples/template-build.ts
/**
 * Template Building Example
 *
 * Shows how to build a custom template and create sandboxes from it.
 */

import { Template, waitForPort, Sandbox } from '../src/index.js';

async function main() {
  console.log('Template Building Example\n');

  // Generate unique template name
  const templateName = `example-python-app-${Date.now()}`;
  console.log(`Template name: ${templateName}\n`);

  // 1. Define a Python web app template
  console.log('1. Defining template...');
  const template = new Template()
    .fromPythonImage('3.11')
    .runCmd('mkdir -p /app')
    .setWorkdir('/app')
    .runCmd(`cat > main.py << 'EOF'
#!/usr/bin/env python3
from http.server import HTTPServer, BaseHTTPRequestHandler
import os

PORT = int(os.environ.get("PORT", 8000))

class SimpleHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(b"<h1>Hello from Hopx!</h1>")

    def log_message(self, format, *args):
        pass

if __name__ == "__main__":
    server = HTTPServer(("0.0.0.0", PORT), SimpleHandler)
    print(f"Server running on port {PORT}")
    server.serve_forever()
EOF`)
    .setEnv('PORT', '8000')
    .setStartCmd('python main.py', waitForPort(8000));

  console.log(`   Template defined with ${template.getSteps().length} build steps`);

  // 2. Build the template
  console.log('\n2. Building template...');
  const result = await Template.build(template, {
    name: templateName,
    apiKey: process.env.HOPX_API_KEY || '',
    baseURL: process.env.HOPX_BASE_URL || 'https://api.hopx.dev',
    cpu: 2,
    memory: 2048,
    diskGB: 10,
    contextPath: process.cwd(),
    onLog: (log) => {
      const level = log.level || 'INFO';
      const message = log.message || '';
      console.log(`   [${level}] ${message}`);
    },
    onProgress: (progress) => {
      console.log(`   Progress: ${progress}%`);
    },
  });

  console.log('\n   Template built successfully');
  console.log(`   Template ID: ${result.templateID}`);
  console.log(`   Build ID: ${result.buildID}`);
  console.log(`   Duration: ${result.duration}ms`);

  // 3. Create sandbox from template
  console.log('\n3. Creating sandbox from template...');
  const sandbox = await Sandbox.create({
    template: templateName,  // Use the template we just built
    envVars: {
      DATABASE_URL: 'postgresql://localhost/mydb',
      API_KEY: 'secret123',
    },
  });

  console.log('   Sandbox created');
  console.log(`   Sandbox ID: ${sandbox.sandboxId}`);

  // 4. Test the sandbox
  console.log('\n4. Testing sandbox...');
  const result2 = await sandbox.runCode('print("Template is working!")', { language: 'python' });
  console.log(`   Output: ${result2.stdout?.trim()}`);

  // 5. Get preview URL
  const previewUrl = await sandbox.getPreviewUrl(8000);
  console.log(`   Preview URL: ${previewUrl}`);

  // 6. Cleanup
  console.log('\n5. Cleaning up...');
  await sandbox.kill();
  console.log('   Sandbox killed');

  console.log('\nDone');
}

// Run example
main().catch(error => {
  console.error('Error:', error.message);
  process.exit(1);
});

File: python/examples/template_build.py
#!/usr/bin/env python3
"""
Template Building Example

Shows how to build a custom template and create VMs from it.
"""

import os
import asyncio
from hopx_ai import Template, wait_for_port


async def main():
    print("🚀 Template Building Example\n")
    
    # 1. Define a Python web app template
    print("1. Defining template...")
    template = (
        Template()
        .from_python_image("3.11")
        .copy("app/", "/app/")
        .pip_install()
        .set_env("PORT", "8000")
        .set_start_cmd("python /app/main.py", wait_for_port(8000))
    )
    
    print(f"   ✅ Template defined with {len(template.get_steps())} steps")
    
    # 2. Build the template
    print("\n2. Building template...")
    
    from hopx_ai.template import BuildOptions
    
    result = await Template.build(
        template,
        BuildOptions(
            name="my-python-app",
            api_key=os.environ.get("HOPX_API_KEY", ""),
            base_url=os.environ.get("HOPX_BASE_URL", "https://api.hopx.dev"),
            cpu=2,
            memory=2048,
            disk_gb=10,
            context_path=os.getcwd(),
            on_log=lambda log: print(f"   [{log['level']}] {log['message']}"),
            on_progress=lambda progress: print(f"   Progress: {progress}%"),
        ),
    )
    
    print("\n   ✅ Template built successfully!")
    print(f"   Template ID: {result.template_id}")
    print(f"   Build ID: {result.build_id}")
    print(f"   Duration: {result.duration}ms")
    
    # 3. Create sandbox from template
    print("\n3. Creating sandbox from template...")

    from hopx_ai import AsyncSandbox

    sandbox = await AsyncSandbox.create(
        template="my-python-app",  # Use template name/alias
        env_vars={
            "DATABASE_URL": "postgresql://localhost/mydb",
            "API_KEY": "secret123",
        },
    )

    print("   ✅ Sandbox created!")
    print(f"   Sandbox ID: {sandbox.sandbox_id}")
    info = await sandbox.get_info()
    print(f"   Status: {info.status}")
    print(f"   Agent URL: {await sandbox.agent_url}")

    # 4. Use the sandbox
    print("\n4. Testing sandbox...")
    result = await sandbox.run_code("""
import os
print(f"DATABASE_URL: {os.environ.get('DATABASE_URL')}")
print(f"API_KEY: {'*' * len(os.environ.get('API_KEY', ''))}")
print("Web app is running on port 8000!")
""", language="python")

    print(f"   Output:\n{result.stdout}")

    # 5. Cleanup
    print("\n5. Cleaning up...")
    await sandbox.kill()
    print("   ✅ Sandbox destroyed")
    
    print("\n✨ Done!")


if __name__ == "__main__":
    asyncio.run(main())

